### PR TITLE
Separated the vectorization mechanism from the Factorization class

### PR DESCRIPTION
**Pull request by yhtang**
> Yu-Hang  1 hour ago
> I am thinking about making the Factorization class agnostic of vectorization. This should remove many complications surround what if nvec=1.
> 
> Yu-Hang  1 hour ago
> The factorize method will carry out vectorization regardless, though.
> 
> Yu-Hang  1 hour ago
> However, vectorization will not be done at the initialization of a Factorization object. It will be done by a pair of standalone vectorize/view/devectorize methods, kind of like what JAX does.
> 
> Daan Camps  43 minutes ago
> That sounds like a neat idea. Probably the vectorize and view methods are sufficient? A devectorize is not really required if we have a view
>

### CODE REVIEW

File: funfact/algorithm.py [hunk 1]
```diff
@@ -102,22 +132,28 @@ def _loss(fac, target):
                 new_best.append(b)
             best_fac.factors = new_best
 
-            if stop_by == 'best':
+            if stop_by == 'first':
                 if np.sum(converged) > 1:
```
Comment (campsd) [inline funfact/algorithm.py:136]: I just realize that we can stop already if `np.sum(converged) > 0` in this case

File: funfact/algorithm.py [hunk 2]
```diff
@@ -102,22 +132,28 @@ def _loss(fac, target):
...
                     pbar.update(max_steps - step)
                     break
             elif isinstance(stop_by, int):
                 if np.sum(converged) > stop_by:
```
Comment (campsd) [inline funfact/algorithm.py:140]: Same here `np.sum(converged) > stop_by - 1`

**Review by campsd — APPROVED** [review]
Looks good to me.

File: funfact/algorithm.py (hunk 1 above)
Comment (yhtang) [reply funfact/algorithm.py:136]: Good point. I'll change that.

**Pull request merged**